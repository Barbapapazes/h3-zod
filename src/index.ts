import type { CompatibilityEvent } from 'h3'
import { createError, useBody, useQuery } from 'h3'
import { z } from 'zod'

// copy of the private Zod utility type of ZodObject
type UnknownKeysParam = 'passthrough' | 'strict' | 'strip'

type TQuery<U extends UnknownKeysParam = any> =
  | z.ZodObject<any, U>
  | z.ZodUnion<[TQuery<U>, ...TQuery<U>[]]>
  | z.ZodIntersection<TQuery<U>, TQuery<U>>
  | z.ZodDiscriminatedUnion<string, z.Primitive, z.ZodObject<any, U>>

export function useValidatedQuery<T extends TQuery>(
  event: CompatibilityEvent,
  schema: T,
) {
  const query = useQuery(event)
  const parsed = schema.safeParse(query)

  if (!parsed.success) {
    throw createError({
      statusCode: 400,
      statusMessage: JSON.stringify(parsed.error.errors),
    })
  }

  return parsed.data as z.infer<T>
}

export async function useValidatedBody<T extends TQuery>(
  event: CompatibilityEvent,
  schema: T,
) {
  const body = await useBody(event)
  const parsed = schema.safeParse(body)

  if (!parsed.success) {
    throw createError({
      statusCode: 400,
      statusMessage: JSON.stringify(parsed.error.errors),
    })
  }

  return parsed.data as z.infer<T>
}

export {
  z,
}
