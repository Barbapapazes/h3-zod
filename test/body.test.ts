import type { SuperTest, Test } from 'supertest'
import supertest from 'supertest'
import { beforeEach, describe, expect, it } from 'vitest'
import type { App, CompatibilityEvent } from 'h3'
import { createApp } from 'h3'
import { useValidatedBody, z } from '../src'

describe('useValidatedBody', () => {
  let app: App
  let request: SuperTest<Test>

  beforeEach(() => {
    app = createApp({ debug: false })
    request = supertest(app)
  })

  const bodySchema = z.object({
    optional: z.string().optional(),
    required: z.boolean(),
  })

  it('returns 200 OK if body matches validation schema', async () => {
    app.use('/validate', async (req: CompatibilityEvent) => await useValidatedBody(req, bodySchema))

    const res = await request.post('/validate').send({ required: true })

    expect(res.status).toEqual(200)
  })

  it('throws 400 Bad Request if body does not match validation schema', async () => {
    app.use('/validate', async (req: CompatibilityEvent) => await useValidatedBody(req, bodySchema))

    const res = await request.post('/validate').send({})

    expect(res.status).toEqual(400)
    expect(res.body).toMatchInlineSnapshot(`{
      "stack": [],
      "statusCode": 400,
      "statusMessage": "{\\"errors\\":[{\\"code\\":\\"invalid_type\\",\\"expected\\":\\"boolean\\",\\"received\\":\\"undefined\\",\\"path\\":[\\"required\\"],\\"message\\":\\"Required\\"}]}",
    }`)
  })
})
