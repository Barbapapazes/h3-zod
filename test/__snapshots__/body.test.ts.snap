// Vitest Snapshot v1

exports[`useValidatedBody > throws 400 Bad Request if body does not match validation schema 1`] = `
{
  "stack": [],
  "statusCode": 400,
  "statusMessage": "[{\\"code\\":\\"invalid_type\\",\\"expected\\":\\"boolean\\",\\"received\\":\\"undefined\\",\\"path\\":[\\"required\\"],\\"message\\":\\"Required\\"}]",
}
`;
